torch/csrc/autograd/profiler_kineto.cpp

```diff
@@ -0,0 +1,298 @@
+        }
+
+        ctx_ptr->sequenceNr = fn.seqNr();
+        ctx_ptr->fwdThreadId = fn.forwardThreadId();
+        ctx_ptr->recFunScope = (uint8_t)fn.scope();
+
+#ifndef C10_MOBILE
+        // backward nodes source range corresponds to the forward node
+        // TODO: consider using C++ stack trace
+        if (state_ptr->config().with_stack &&
+            fn.scope() != at::RecordScope::BACKWARD_FUNCTION) {
+          auto cs = prepareCallstack(jit::currentCallstack());
+          if (cs.empty()) {
+            cs = prepareCallstack(jit::tracer::pythonCallstack());
+          }
+          ctx_ptr->stack = callstackStr(cs);
+        }
+#endif
+        return ctx_ptr;
+      },
+      [](const at::RecordFunction& fn, at::ObserverContext* ctx_ptr) {
+        auto state_ptr = getProfilerTLSState();
+        if (!state_ptr || state_ptr->config().state != ProfilerState::KINETO) {
+          return;
+        }
+        auto kineto_ctx_ptr = dynamic_cast<KinetoObserverContext*>(ctx_ptr);
+        TORCH_INTERNAL_ASSERT(kineto_ctx_ptr != nullptr);
+
+        kineto_ctx_ptr->endThreadId = at::RecordFunction::currentThreadId();
+
+        state_ptr->reportClientActivity(fn, kineto_ctx_ptr);
+        libkineto::api().popCorrelationId();
+      })
+    .needsInputs(state_ptr->config().report_input_shapes)
+    .needsIds(true));
+  state_ptr->setCallbackHandle(handle);
+}
+
+std::string shapesToStr(const std::vector<std::vector<int64_t>>& shapes) {
+  std::ostringstream oss;
+  oss << "[";
+  for (auto t_idx = 0; t_idx < shapes.size(); ++t_idx) {
+    if (t_idx > 0) {
+      oss << ", ";
+    }
+    oss << "[";
+    for (auto s_idx = 0; s_idx < shapes[t_idx].size(); ++s_idx) {
+      if (s_idx > 0) {
+        oss << ", ";
+      }
+      oss << shapes[t_idx][s_idx];
+    }
+    oss << "]";
+  }
+  oss << "]";
+  return oss.str();
+}
+
+} // namespace
+
+void prepareProfiler(
+    const ProfilerConfig& config,
+    const std::set<ActivityType>& activities) {
+  TORCH_CHECK(config.state == ProfilerState::KINETO,
+      "Supported only in Kineto profiler");
+
+  std::set<libkineto::ActivityType> k_activities;
+  if (activities.count(ActivityType::CPU)) {
+    k_activities.insert(libkineto::ActivityType::EXTERNAL_CORRELATION);
+    k_activities.insert(libkineto::ActivityType::CUDA_RUNTIME);
+  }
+  //if (activities.count(ActivityType::CUDA_RUNTIME)) {
+  //  k_activities.insert(libkineto::ActivityType::CUDA_RUNTIME);
+  //}
+  if (activities.count(ActivityType::CUDA)) {
+    k_activities.insert(libkineto::ActivityType::GPU_MEMCPY);
+    k_activities.insert(libkineto::ActivityType::GPU_MEMSET);
+    k_activities.insert(libkineto::ActivityType::CONCURRENT_KERNEL);
+    k_activities.insert(libkineto::ActivityType::CUDA_RUNTIME);
+  }
+
+  //if (!libkineto::api().hasProfilerRegistered()) {
+  //  libkineto::api().registerProfiler(
+  //    std::make_unique<libkineto::ActivityProfilerInterface>(false));
+  //}
+  if (!libkineto::api().isProfilerInitialized()) {
```

Comment:
As discussed we could move this inside prepareTrace() to hide it away (in a follow-up).